src/core/server/saved_objects/service/lib/aggs_utils.ts

```diff
@@ -0,0 +1,58 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { IndexMapping } from '../../mappings';
+
+import { SavedObjectsErrorHelpers } from './errors';
+import { hasFilterKeyError } from './filter_utils';
+import { SavedObjectAggs, validateSavedObjectTypeAggs } from './saved_object_aggs_types';
+
+export const validateGetSavedObjectAggs = (
+  allowedTypes: string[],
+  aggs: SavedObjectAggs,
+  indexMapping: IndexMapping
+): SavedObjectAggs => {
+  validateSavedObjectTypeAggs(aggs);
+  return validateGetAggFieldValue(allowedTypes, aggs, indexMapping);
+};
+
+const validateGetAggFieldValue = (
+  allowedTypes: string[],
+  aggs: any,
+  indexMapping: IndexMapping
+): SavedObjectAggs => {
+  return Object.keys(aggs).reduce((acc, key) => {
+    if (key === 'field') {
```

Comment:
Aggregations seem to have different conventions for specifying the fields on which the operate for instance see the example for https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-adjacency-matrix-aggregation.html#search-aggregations-bucket-adjacency-matrix-aggregation

So it looks like we will have run `hasFilterKeyError` on different keys depending on the aggregation type, like `aggs.*.adjacency_matrix.*.filters.*.terms.*`. Ideally, this `hasFilterKeyError` validation for each aggregation type will live alongside the io-ts type for that aggregation.

 I'm not sure if there's a way to inject the `allowedTypes` and `indexMapping` state into an io-ts validation, otherwise we might have to come up with another architecture to compose validations. It's not strictly necessary, but would be nice if we didn't need to recurse the aggregation twice and raise different errors for different validation failures, but could instead use one approach to find both schema errors and invalid field names or invalid saved object type errors.

Reply:
`hasFilterKeyError` Makes sure that fields are always valid, i.e. they exist on the mappings for the queried type and that makes the developer experience a lot better, but I don't think this is a security risk. If you query the `dashboard` type but specify an aggregation on `vizualization.attributes.foo` you'll just not get any results, this field will always be undefined.

But we still have to do the `aggs[key].replace('.attributes', '')` part if we want to keep the KQL convention of making a distinction between type properties and saved object properties.

Reply:
I think that you can re-write the aggregation differently to pass our test for example:
```
GET emails/_search
{
  "size": 0,
  "aggs" : {
    "interactions" : {
      "adjacency_matrix" : {
        "filters" : {
          "grpA" : { "terms" : { "accounts" : ["hillary", "sidney"] }},
          "grpB" : { "terms" : { "accounts" : ["donald", "mitt"] }},
          "grpC" : { "terms" : { "accounts" : ["vladimir", "nigel"] }}
        }
      }
    }
  }
}
```
can become this and it work the same way
```
GET emails/_search
{
  "size": 0,
  "aggs" : {
    "interactions" : {
      "adjacency_matrix" : {
        "filters" : {
          "grpA" : { "terms" : { "fields":"accounts", "includes": ["hillary", "sidney"] }},
          "grpB" : { "terms" : { "fields": "accounts", "includes": ["donald", "mitt"] }},
          "grpC" : { "terms" : { "fields": "accounts" , "includes": ["vladimir", "nigel"] }}
        }
      }
    }
  }
}
```

Reply:
Why we used the key attributes for the filter, it was for differentiating between the property who belongs to the saved object type and the native property of the saved object. Therefore, I think we should keep it the same way to be concise and identical. It will make more sense to end-user of the API.

Reply:
Ah ok, if we can use an alternative syntax to make the fields explicit that solves this conundrum 😅 We would just have to educate developers that they always need to use this alternative syntax which might not be obvious, but I don't think that's a problem.